src/perf_counter/metrics.cpp

```diff
@@ -0,0 +1,194 @@
+// under the License.
+
+#include <dsn/metrics.h>
+#include <dsn/utility/singleton.h>
+#include <dsn/utility/smart_pointers.h>
+#include <mutex>
+#include <utility>
+
+// The metric library is designed to entirely replace the "perf_counter"
+// library on the API level.
+// Internally, the implementation of metrics still relies on the
+// existing "perf_counter" code.
+#include "perf_counter/perf_counter_atomic.h"
+#include <dsn/c/api_utilities.h>
+
+namespace dsn {
+
+class metric_registry : public utils::singleton<metric_registry>
+{
+public:
+    void register_entity(const std::string &id, ref_ptr<metric_entity> entity)
+    {
+        std::lock_guard<std::mutex> guard(_mtx);
+        dassert(_entities_map.find(id) == _entities_map.end(),
+                "repeatedly register metric_entity [id=%s]",
+                id.data());
+        _entities_map[id] = std::move(entity);
+    }
+
+    void collect_metrics(metric_data_sink *sink)
+    {
+        std::lock_guard<std::mutex> guard(_mtx);
+        for (auto it = _entities_map.begin(); it != _entities_map.end();) {
+            auto next_it = std::next(it);
+            sink->iterate_entity(*it->second);
+            it->second->collect_metrics(sink);
+            if (it->second->num_metrics() == 0 && it->second->get_count() == 1) {
+                // This entity has no metrics and either has no more external references,
+                // so we can remove it.
+                _entities_map.erase(it);
+            }
+            it = next_it;
+        }
+    }
+
+private:
+    metric_registry() = default;
+    friend class utils::singleton<metric_registry>;
+
+private:
+    mutable std::mutex _mtx;
+    std::unordered_map<std::string, ref_ptr<metric_entity>> _entities_map;
+};
+
+METRIC_DEFINE_entity(server);
+
+/*extern*/ ref_ptr<metric_entity> get_metric_entity_server()
+{
+    static auto server_entity = METRIC_ENTITY_server.instantiate("server", {});
+    return server_entity;
+}
+
+ref_ptr<metric_entity> metric_entity_prototype::instantiate(const std::string &id,
+                                                            metric_entity::attrs_map_t attrs)
+{
+    dassert(attrs.find("entity") == attrs.end(), "attribute \"entity\" is reserved");
+    attrs["entity"] = _name;
+    ref_ptr<metric_entity> entity(new metric_entity(id, std::move(attrs)));
+    metric_registry::instance().register_entity(id, entity);
+    return entity;
+}
+
+void metric_entity::collect_metrics(metric_data_sink *sink)
+{
+    std::lock_guard<std::mutex> guard(_mtx);
+    for (auto it = _metric_map.begin(); it != _metric_map.end();) {
+        auto next_it = std::next(it);
+        if (it->second->get_count() == 1) {
+            _metric_map.erase(it);
+        } else {
+            it->second->take_snapshot(sink);
+        }
+        it = next_it;
+    }
+}
+
+class gauge_t::impl
+{
+public:
+    explicit impl(metric_ctor_args args) : _value("", "", args.name.data(), COUNTER_TYPE_NUMBER, "")
```

Comment:
In the fowllowng refactor, will you remove the leading two paramaeters of perf_counter_number_atomic?